garage/tf/models/base.py

```diff
@@ -0,0 +1,368 @@
+        # Here we pass the output from the above SimpleModel object.
+        model_2 = ComplexModel()
+        output_2 = model_2.build(output)
+
+    """
+
+    def build(self, *inputs):
+        """
+        Output of model with the given input placeholder(s).
+
+        This function is implemented by subclasses to create their computation
+        graphs, which will be managed by Model. Generally, subclasses should
+        implement `build()` directly.
+        """
+        pass
+
+    @property
+    def name(self):
+        """Name for this Model."""
+        pass
+
+    @property
+    def input(self):
+        """
+        Tensor input of the Model.
+
+        The input can be either a placeholder, or the output from another
+        model.
+        """
+        pass
+
+    @property
+    def output(self):
+        """
+        Tensor output of the Model.
+
+        The output can be passed to another model as an input.
+        """
+        pass
+
+    @property
+    def inputs(self):
+        """
+        Tensor inputs of the Model.
+
+        The inputs can be either placeholders, or the outputs from another
+        model.
+        """
+        pass
+
+    @property
+    def outputs(self):
+        """
+        Tensor outputs of the Model.
+
+        The outputs can be passed to another model as an input.
+        """
+        pass
+
+    @property
+    def parameters(self):
+        """
+        Parameters of the Model.
+
+        The output of a model is determined by its parameter. It could be
+        the weights of a neural network model or parameters of a loss
+        function model.
+        """
+        pass
+
+    @parameters.setter
+    def parameters(self, parameters):
+        """Set parameters of the Model."""
+        pass
+
+
+class Model(BaseModel, metaclass=abc.ABCMeta):
+    """
+    Abstract class for models with automatic pickling.
+
+    This class follows the BaseModel API and handles pickling automatically.
+
+    The target weights should be assigned to self._default_parameters before
+    pickling, so that the newly created model can check if target weights exist
+    or not. When unpickled, the unserialized model will load the weights
+    from self._default_parameters.
+    """
+
+    def __init__(self):
+        self._name = type(self).__name__  # name default to class
+        self._inputs = None
+        self._outputs = None
+        self._default_parameters = None
+
+    @property
+    def name(self):
+        """Name of the model."""
+        return self._name
+
+    @property
+    def parameters(self):
+        """Parameters of the Model."""
+        raise NotImplementedError
+
+    @parameters.setter
+    def parameters(self, parameters):
+        """Set model parameters."""
+        raise NotImplementedError
+
+    def __getstate__(self):
+        """Object.__getstate__."""
+        new_dict = self.__dict__.copy()
+        del new_dict['_networks']
```

Comment:
Wait, so `Model` knows about the `_networks` field, but doesn't actually have it? That's weird. We should just merge `TfModel` into `Model`, since it's in the `tf` directory already.

Reply:
agreed

Reply:
We have to remove the `networks` when pickling. The networks are connectivity and I think they can be removed.